gym_graph_traffic/envs/segment.py

```diff
@@ -93,23 +93,41 @@ def update_first_phase(self) -> None:
         First phase of segment update: cellular automata step, and (sometimes) passing car to following segment.
         """
 
-        # extend p vector by free cells of following segment
-        next_segment_free_cells = self.next_intersection.can_i_go(self.idx)
-        if next_segment_free_cells > 0:
-            self.p = np.append(self.p, np.zeros(next_segment_free_cells))
-
-        # update cellular automata
-        self._nagel_schreckenberg_step()
+        info_can_i_go = None
+        # check if there is auto (1) in the last 5 cells of the segment
+        if 1 in self.p[-5:]:
+            # get car indices from the last cells
+            cars_indices = [i for i in self.p.nonzero()[0] if i >= 95]
+            # Check if the car can pass the intersection, if so then save all the informations into the dictionary
+            # info_can_i_go =
+            # {'free_cells_at_intersection': 0 or 1 or 2 or 3,
+            # 'chosen_segment':[intersection.exits[chosen_direction].idx),intersection.exits[chosen_direction].to_side],
+            # 'free_cells_at_segment': {cars_indices[-1]: intersection.exits[chosen_direction].free_init_cells}},
+            # 'direction': "straight" or "turn right" or "turn left" }
+            # Otherwise info_can_i_go = 0, so car can't go through the intersection.
+            info_can_i_go = self.next_intersection.can_i_go(self.idx, cars_indices[-1])
+            if info_can_i_go is not None and isinstance(info_can_i_go, dict):
+                if info_can_i_go.get("free_cells_at_intersection") != 0:
+                    # extend p vector by free cells of intersection and following segment
+                    self.p = np.append(self.p, np.zeros(info_can_i_go['free_cells_at_segment'][cars_indices[-1]] +
+                                                        info_can_i_go['free_cells_at_intersection']))
+
+        # update cellular automata if vector p contains cars
+        if 1 in self.p:
+            self._nagel_schreckenberg_step()
 
         # cut excessive cells
         self.p, next_segment_cells = np.split(self.p, [self.length])
 
         # if some car crossed intersection: pass car to next segment (via the intersection)
         try:
             next_sect_car_pos = next_segment_cells.tolist().index(1)
-            # at any given update, only one car can cross intersection (by the rules of automata)
+            # at any given update, only two cars can cross intersection (by the rules of automata)
```

Comment:
Both new and previous assumptions seem kind of ad-hoc to me. It would be good to decouple Segments implementation from Intersection's, or at least parameterize it somehow.